src/main/java/graphql/nadel/engine/OverallQueryTransformer.java

```diff
@@ -282,20 +278,29 @@ QueryTransformationResult transformMergedFields(
                                                                Map<String, Object> variableValues,
                                                                Service service,
                                                                Object serviceContext) {
-        return fragments.values().stream()
-                .map(fragment -> transformFragmentDefinition(
-                        executionContext,
-                        underlyingSchema,
-                        fragment,
-                        transformationByResultField,
-                        typeRenameMappings,
-                        referencedFragmentNames,
-                        referencedVariables,
-                        serviceExecutionHooks,
-                        variableValues,
-                        service,
-                        serviceContext))
-                .collect(toMapCollector(FragmentDefinition::getName, identity()));
+
+        Set<String> fragmentsToTransform = new LinkedHashSet<>(referencedFragmentNames);
```

Comment:
Is that a Deque if you are popping off elements rather than a Set with remove

Reply:
its a Set: we don't want to transform fragments twice